Zotero/Scenes/Detail/PDF/Views/PdfThumbnailsViewController.swift

```diff
@@ -0,0 +1,133 @@
+        setupCollectionView()
+
+        viewModel.stateObservable
+            .observe(on: MainScheduler.instance)
+            .subscribe(with: self, onNext: { `self`, state in
+                self.update(state: state)
+            })
+            .disposed(by: disposeBag)
+
+        func setupCollectionView() {
+            let registration = cellRegistration
+            dataSource = UICollectionViewDiffableDataSource(collectionView: collectionView, cellProvider: { collectionView, indexPath, itemIdentifier in
+                return collectionView.dequeueConfiguredReusableCell(using: registration, for: indexPath, item: itemIdentifier)
+            })
+            collectionView.dataSource = dataSource
+            collectionView.delegate = self
+            collectionView.prefetchDataSource = self
+            collectionView.backgroundColor = .systemGray6
+        }
+    }
+
+    override func viewIsAppearing(_ animated: Bool) {
+        super.viewIsAppearing(animated)
+        viewModel.process(action: .loadPages)
+    }
+
+    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
+        super.traitCollectionDidChange(previousTraitCollection)
+        guard traitCollection.hasDifferentColorAppearance(comparedTo: previousTraitCollection) else { return }
+        self.viewModel.process(action: .setUserInterface(isDark: traitCollection.userInterfaceStyle == .dark))
+    }
+
+    func set(visiblePage: Int) {
+        viewModel.process(action: .setSelectedPage(pageIndex: visiblePage, type: .fromDocument))
+    }
+
+    private func update(state: PdfThumbnailsState) {
+        if state.changes.contains(.pages) {
+            var snapshot = NSDiffableDataSourceSnapshot<Int, String>()
+            snapshot.appendSections([0])
+            snapshot.appendItems(state.pages)
+            dataSource.apply(snapshot) { [weak self] in
+                guard let self else { return }
+                let indexPath = IndexPath(row: self.viewModel.state.selectedPageIndex, section: 0)
+                self.collectionView.selectItem(at: indexPath, animated: false, scrollPosition: .centeredVertically)
+            }
+        }
+
+        if let index = state.loadedThumbnail, index < dataSource.snapshot().itemIdentifiers.count {
+            var snapshot = dataSource.snapshot()
+            let label = dataSource.snapshot().itemIdentifiers[index]
+            snapshot.reloadItems([label])
+            dataSource.apply(snapshot)
+        }
+
+        if state.changes.contains(.userInterface) {
+            var snapshot = dataSource.snapshot()
+            snapshot.reloadSections([0])
+            dataSource.apply(snapshot)
+        }
+
+        if state.changes.contains(.scrollToSelection) {
+            let indexPath = IndexPath(row: state.selectedPageIndex, section: 0)
+            collectionView.selectItem(at: indexPath, animated: true, scrollPosition: .centeredVertically)
+        }
```

Comment:
I got a crash when restoring a document at launch, but could be a race issue in general.

```suggestion
        // The following updates should be ignored if the collection hasn't loaded yet for the first time.
        var snapshot = dataSource.snapshot()
        guard snapshot.numberOfSections > 0 else { return }

        if state.changes.contains(.userInterface) {
            snapshot.reloadSections([0])
            dataSource.apply(snapshot)
        }

        if state.changes.contains(.scrollToSelection) {
            let indexPath = IndexPath(row: state.selectedPageIndex, section: 0)
            collectionView.selectItem(at: indexPath, animated: true, scrollPosition: .centeredVertically)
        }
```

Reply:
Yes, indeed that was what caused the crash, as it caused an update.